tests/test_version_traversal.py

```diff
@@ -308,15 +313,139 @@ def test_relative_upgrade(self):
             set([self.c2.revision]),
         )
 
-    def test_relative_downgrade(self):
+    def test_relative_downgrade_baseplus2(self):
+        """ base+2 points to b, no branch label, drop everything above b. """
+        self._assert_downgrade(
+            "base+2",
+            [self.d2.revision, self.d1.revision],
+            [
+                self.down_(self.d2),
+                self.down_(self.c2),
+                self.down_(self.d1),
+                self.down_(self.c1),
+            ],
+            set([self.b.revision]),
+        )
 
+    def test_relative_downgrade_branchplus2(self):
+        """
+        Correct behaviour (per
+        https://github.com/sqlalchemy/alembic/pull/763#issuecomment-738741297)
+        Only the c2branch should be downgraded, right back to base+2 = b
+        """
         self._assert_downgrade(
             "c2branch@base+2",
             [self.d2.revision, self.d1.revision],
-            [self.down_(self.d2), self.down_(self.c2), self.down_(self.d1)],
+            [
+                self.down_(self.d2),
+                self.down_(self.c2),
+            ],
+            set([self.d1.revision]),
+        )
+
+    def test_relative_downgrade_branchplus3(self):
+        """ c2branch@base+3 equivalent to c2. """
+        self._assert_downgrade(
+            self.c2.revision,
+            [self.d2.revision, self.d1.revision],
+            [self.down_(self.d2)],
+            set([self.d1.revision, self.c2.revision]),
+        )
+        self._assert_downgrade(
+            "c2branch@base+3",
+            [self.d2.revision, self.d1.revision],
+            [self.down_(self.d2)],
+            set([self.d1.revision, self.c2.revision]),
+        )
+
+    # Old downgrade -1 behaviour depends on order of branch upgrades.
+    # This should probably fail (ambiguous) but is currently documented
```

Comment:
**mike bayer** (zzzeek) wrote:

OK, I would assume this refers to the "downgrade -1" example at https://alembic.sqlalchemy.org/en/latest/branches.html.   So, can I assume that prior to this behavior, the example in the docs is basically random, right?   OK.   I see the failure here looks like this:

    File "/home/classic/dev/alembic/alembic/script/revision.py", line 1005, in _parse_downgrade
        assert len(uppers) == 1

OK great.  So, because we are at 1.5.2 / 1.5.3 etc. for this, we can't rip out that old behavior entirely right now.  Can we have your assertion run, and if it fails, emit a warning, that this case is deprecated, then proceed to have it do the old thing it was doing?   then we also need to update the docs to illustrate the "-1" thing using a proper branch label.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2530

Reply:
**Simon Bowly** (simonbowly) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2530